clang/lib/Sema/SemaDeclCXX.cpp

```diff
@@ -9108,7 +9123,8 @@ bool Sema::CheckExplicitlyDefaultedComparison(Scope *S, FunctionDecl *FD,
   //   declaration, it is implicitly considered to be constexpr.
   // FIXME: Only applying this to the first declaration seems problematic, as
   // simple reorderings can affect the meaning of the program.
-  if (First && !FD->isConstexpr() && Info.Constexpr)
+  if ((First && !FD->isConstexpr() && Info.Constexpr) ||
+      getLangOpts().CPlusPlus23)
```

Comment:
This change perhaps needs a test.

Reply:
isn't `Info.Constexpr` always true in c++ mode?
I'm trying to think of a meaningful test.

Haybe calling a comparison function of a class with a non-literal member from a consteval function?

Reply:
> isn't Info.Constexpr always true in c++ mode?

Hmm, then we probably don't need this change at all?

> Haybe calling a comparison function of a class with a non-literal member from a consteval function?

Should the call to such function be allowed? I suppose not due to-literal member. The problem is, there result is same with our without this change. I meant to apply this:

> CheckExplicitlyDefaultedComparison should alwways set the constexpr flag in c++23 mode per https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2448r2.html#pnum_39

Did I misunderstand?

Reply:
> isn't Info.Constexpr always true in c++ mode?

**C++23**, sorry 

> Should the call to such function be allowed? I suppose not due to-literal member. The problem is, there result is same with or without this change. I meant to apply this

Hum right, i was thinking we could have a `if(false)` in the comparison function, but that's for defaulted function so it doesn't work.

The error message should be different though, right? something like "the call to operator== is not a constant expression" ?

Reply:
> isn't Info.Constexpr always true in c++ mode?
C++23, sorry

Not sure it is always true even for C++23, I think it is true when the operator produces constexpr. Otherwise we would have never emitted warn_cxx23_compat_defaulted_comparison_constexpr_mismatch a bit earlier than my change, and we do.

> The error message should be different though, right? something like "the call to operator== is not a constant expression" ?

In both cases it says "non-literal type 'ComparisonNonConstexpr' cannot be used in a constant expression".

Reply:
I'm not sure actually that this change is required.
If the function is marked constexpr then we're all set here.
If the function is not explicitly marked constexpr, we could force it constexpr for c++23, but it is already done if it is Info.Constexpr is true. 
If Info.Constexpr is false it means that the function is constexpr only in c++23 and it means it can't be called from real constexpr
context anyway, additionally marking it constexpr doesn't lift more errors.
The point of p2448r2 here is that we can attach constexpr explicitly to it.